File: src/phase4/partition.ts
/**
 * Phase 4 — physical-domain partitioning + pre-dispatch guards (§4, §5, §6).
 *
 * Domain boundary = CONNECTED COMPONENT of the physical seat graph (faithful
 * mirror of the frozen Python partitioner, partition.py). Hall names are NEVER
 * used as proof of independence: if a configured cross-hall adjacency edge
 * connects two halls, they belong to the same solver domain. The three
 * partition invariants are computed and verified explicitly.
 *
 * Guards mirror the FROZEN Python guard semantics (guards.py): risk ratios are
 * configurable scheduling signals, NOT infeasibility proofs; the only true
 * capacity error is candidateCount > seatCount.
 */
import { GUARD_ERR_CODES, MAX_DOMAIN_CANDIDATES } from "./types";
import type {
  DomainCandidate,
  DomainGuardResult,
  DomainHall,
  DomainPlan,
  GuardClassification,
} from "./types";
import {
  buildPhysicalGraph,
  connectedComponents,
  verifyPartitionInvariants,
  topologyAnomalyEvidence,
  type AdjacencyMode,
  type CrossHallEdgeInput,
} from "./topology";

export interface CompositionLimits {
  largestDepartmentRatio: number;
  largestYearRatio: number;
  largestCohortRatio: number;
  maxEmptySeatRatio: number;
  /** Topology ceiling (§24) — maximum candidates per connected component. */
  maxDomainCandidates?: number;
}

export const DEFAULT_COMPOSITION_LIMITS: CompositionLimits = {
  largestDepartmentRatio: 0.6,
  largestYearRatio: 0.7,
  largestCohortRatio: 0.5,
  maxEmptySeatRatio: 0.8,
  maxDomainCandidates: MAX_DOMAIN_CANDIDATES,
};

export interface PartitionInput {
  candidates: DomainCandidate[];
  halls: DomainHall[];
  adjacency?: AdjacencyMode;
  crossHallEdges?: CrossHallEdgeInput[];
}

export interface PartitionOutcome {
  domains: DomainPlan[];
  /** True when the session cannot be seated as requested (STOP signal). */
  blocked: boolean;
  blockedReason: string | null;
  unallocatedCount: number;
  /** Any partition-invariant violation (every seat/edge in exactly one domain). */
  invariantErrors: string[];
}

export interface ComponentTopology {
  graphNodes: number;
  graphEdges: number;
  crossHallEdgeCount: number;
  anomalyEvidence: ReturnType<typeof topologyAnomalyEvidence>;
}

/**
 * Balanced allocation: candidates are allocated round-robin across domains
 * (per department, register order) respecting seat capacity — mirroring the
 * frozen allocate_candidates_to_domains. Every domain receives a balanced mix.
 */
export function partitionCandidates(
  input: PartitionInput,
  limits: CompositionLimits = DEFAULT_COMPOSITION_LIMITS,
  maxDomainCandidates: number = MAX_DOMAIN_CANDIDATES,
): PartitionOutcome {
  const { candidates, halls } = input;
  const adjacency = input.adjacency ?? "eight";
  const crossHallEdges = input.crossHallEdges ?? [];

  const graph = buildPhysicalGraph(halls, adjacency, crossHallEdges);
  const components = connectedComponents(graph);
  const indexBySeatId = new Map(graph.nodes.map((n) => [n.seatId, n.index]));

  const totalCapacity = halls.reduce((sum, h) => sum + h.capacity, 0);
  if (candidates.length > totalCapacity) {
    return {
      domains: [],
      blocked: true,
      blockedReason: "insufficient aggregate seat capacity",
      unallocatedCount: candidates.length - totalCapacity,
      invariantErrors: [],
    };
  }

  const domains: DomainPlan[] = components.map((seatIndices, i) => {
    const componentHalls = new Map<string, DomainHall>();
    for (const idx of seatIndices) {
      const node = graph.nodes[idx]!;
      const hall = halls.find((h) => h.id === node.hallId);
      if (hall) componentHalls.set(hall.id, hall);
    }
    const hallList = [...componentHalls.values()].sort((a, b) =>
      a.hallNumber.localeCompare(b.hallNumber),
    );
    const domainId = `D-${hallList.map((h) => h.hallNumber).join("+") || `component-${i}`}`;
    return {
      domainId,
      hallIds: hallList.map((h) => h.id),
      hallNumbers: hallList.map((h) => h.hallNumber),
      halls: hallList,
      seats: hallList.flatMap((h) => h.seats),
      candidates: [],
      seatCount: hallList.reduce((sum, h) => sum + h.capacity, 0),
      candidateCount: 0,
      guard: { classification: "BALANCED", riskViolations: [] },
      blocked: false,
      blockedReason: null,
    };
  });

  // Partition invariants (explicit, not assumed): every seat in exactly one
  // domain and no adjacency edge spans two domains.
  const invariantErrors = verifyPartitionInvariants(
    graph,
    domains.map((d) => ({
      domainId: d.domainId,
      seatIndices: d.seats
        .map((s) => indexBySeatId.get(s.id))
        .filter((idx): idx is number => idx !== undefined),
    })),
  );

  // Candidate allocation: per department (sorted), round-robin across domains.
  const byDepartment = new Map<string, DomainCandidate[]>();
  for (const c of candidates) {
    const bucket = byDepartment.get(c.department) ?? [];
    bucket.push(c);
    byDepartment.set(c.department, bucket);
  }

  const capacityByDomain = domains.map((d) => d.seatCount);
  const assigned = domains.map(() => 0);
  const domainCount = domains.length;
  let allocationFailed: string | null = null;

  for (const dept of [...byDepartment.keys()].sort()) {
    let cursor = 0;
    const sorted = byDepartment.get(dept)!.slice().sort((a, b) => {
      const reg = a.registerNumber.localeCompare(b.registerNumber);
      return reg !== 0 ? reg : a.id.localeCompare(b.id);
    });
    candidateLoop: for (const candidate of sorted) {
      for (let attempt = 0; attempt < domainCount; attempt++) {
        const d = cursor % domainCount;
        cursor += 1;
        if (assigned[d]! < capacityByDomain[d]!) {
          domains[d]!.candidates.push(candidate);
          assigned[d] = (assigned[d] ?? 0) + 1;
          continue candidateLoop;
        }
      }
      allocationFailed =
        allocationFailed ?? `no domain has free seat capacity for candidate ${candidate.id}`;
    }
  }

  // Guards (pre-dispatch validation per domain, §6) + topology ceiling (§24).
  let blocked = false;
  let blockedReason: string | null = null;
  for (const domain of domains) {
    domain.candidateCount = domain.candidates.length;
    domain.guard = computeCompositionGuard(domain.candidates, domain.seatCount, limits);
    if (domain.guard.classification === "INSUFFICIENT_CAPACITY") {
      domain.blocked = true;
      domain.blockedReason = GUARD_ERR_CODES.INSUFFICIENT_CAPACITY;
      blocked = true;
      blockedReason = blockedReason ?? `${domain.domainId}: candidates > seats`;
    } else if (domain.candidateCount > (limits.maxDomainCandidates ?? maxDomainCandidates)) {
      domain.blocked = true;
      domain.blockedReason = GUARD_ERR_CODES.TOPOLOGY_OVERSIZED;
      blocked = true;
      blockedReason = blockedReason ?? GUARD_ERR_CODES.TOPOLOGY_OVERSIZED;
    }
  }

  const unallocatedCount = candidates.length - assigned.reduce((a, b) => a + b, 0);
  if (allocationFailed) {
    blocked = true;
    blockedReason = allocationFailed;
  }
  if (invariantErrors.length > 0) {
    blocked = true;
    blockedReason = blockedReason ?? invariantErrors[0]!;
  }

  return { domains, blocked, blockedReason, unallocatedCount, invariantErrors };
}

/** Faithful mirror of the frozen Python composition guard (guards.py). */
export function computeCompositionGuard(
  candidates: DomainCandidate[],
  seatCount: number,
  limits: CompositionLimits = DEFAULT_COMPOSITION_LIMITS,
): DomainGuardResult {
  const candidateCount = candidates.length;
  if (candidateCount > seatCount) {
    return {
      classification: "INSUFFICIENT_CAPACITY",
      riskViolations: [
        `${GUARD_ERR_CODES.INSUFFICIENT_CAPACITY}: candidates=${candidateCount} > seats=${seatCount}`,
      ],
    };
  }

  const departmentCounts = new Map<string, number>();
  const yearCounts = new Map<string, number>();
  const cohortCounts = new Map<string, number>();
  for (const c of candidates) {
    departmentCounts.set(c.department, (departmentCounts.get(c.department) ?? 0) + 1);
    const year = c.year ?? "<none>";
    yearCounts.set(year, (yearCounts.get(year) ?? 0) + 1);
    const cohort = `${c.department}/${year}`;
    cohortCounts.set(cohort, (cohortCounts.get(cohort) ?? 0) + 1);
  }

  const largestDepartmentCount = maxOf(departmentCounts.values());
  const largestYearCount = maxOf(yearCounts.values());
  const largestCohortCount = maxOf(cohortCounts.values());

  const largestDepartmentRatio = candidateCount ? largestDepartmentCount / candidateCount : 0;
  const largestYearRatio = candidateCount ? largestYearCount / candidateCount : 0;
  const largestCohortRatio = candidateCount ? largestCohortCount / candidateCount : 0;
  const emptySeatRatio = seatCount ? (seatCount - candidateCount) / seatCount : 0;

  const riskViolations: string[] = [];
  if (largestDepartmentRatio > limits.largestDepartmentRatio) {
    riskViolations.push(
      `largestDepartmentRatio=${largestDepartmentRatio.toFixed(2)} > limit ${limits.largestDepartmentRatio}`,
    );
  }
  if (yearCounts.size >= 2 && largestYearRatio > limits.largestYearRatio) {
    riskViolations.push(
      `largestYearRatio=${largestYearRatio.toFixed(2)} > limit ${limits.largestYearRatio}`,
    );
  }
  if (cohortCounts.size >= 2 && largestCohortRatio > limits.largestCohortRatio) {
    riskViolations.push(
      `largestCohortRatio=${largestCohortRatio.toFixed(2)} > limit ${limits.largestCohortRatio}`,
    );
  }
  if (emptySeatRatio > limits.maxEmptySeatRatio) {
    riskViolations.push(`emptySeatRatio=${emptySeatRatio.toFixed(2)} > limit ${limits.maxEmptySeatRatio}`);
  }

  const classification: GuardClassification =
    riskViolations.length === 0 ? "BALANCED" : "IMBALANCE_RISK";
  return { classification, riskViolations };
}

function maxOf(values: Iterable<number>): number {
  let max = 0;
  for (const v of values) max = Math.max(max, v);
  return max;
}

export function componentTopology(
  domain: DomainPlan,
  adjacency: AdjacencyMode = "eight",
  crossHallEdges: CrossHallEdgeInput[] = [],
): ComponentTopology {
  const graph = buildPhysicalGraph(domain.halls, adjacency, crossHallEdges);
  return {
    graphNodes: graph.nodes.length,
    graphEdges: graph.edges.length,
    crossHallEdgeCount: graph.crossHallEdges.length,
    anomalyEvidence: topologyAnomalyEvidence(
      graph,
      domain.seats.map((s) => graph.nodes.findIndex((n) => n.seatId === s.id)),
      domain.domainId,
      domain.candidateCount,
      domain.seatCount,
      MAX_DOMAIN_CANDIDATES,
      domain.hallIds,
    ),
  };
}

File: solver-service/app/partition.py
"""Phase B — Dynamic domain partitioner (§22, §23, §38).

Computes connected components of the physical interaction graph. Each connected
component is a candidate solver domain. The partitioner enforces the three
partition invariants (§9, §10, §38):

- every candidate -> exactly one domain
- every seat -> exactly one domain
- every adjacency edge -> one domain

and provides the topology ceiling guard (§24) with topology anomaly evidence
(§25). An oversized domain is rejected before CP-SAT is ever instantiated.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Optional

from .graph import PhysicalSeatGraph
from .models import Candidate

MAX_DOMAIN_CANDIDATES_DEFAULT = 1000

ERR_GRAPH_TOPOLOGY_OVERSIZED_COMPONENT = "ERR_GRAPH_TOPOLOGY_OVERSIZED_COMPONENT"
ERR_INVALID_DOMAIN_ASSIGNMENT = "ERR_INVALID_DOMAIN_ASSIGNMENT"


@dataclass
class Domain:
    domain_id: str
    seat_indices: list[int] = field(default_factory=list)
    hall_ids: list[str] = field(default_factory=list)
    building_ids: list[str] = field(default_factory=list)
    candidate_ids: list[str] = field(default_factory=list)

    @property
    def seat_count(self) -> int:
        return len(self.seat_indices)

    @property
    def candidate_count(self) -> int:
        return len(self.candidate_ids)


@dataclass
class PartitionResult:
    domains: list[Domain]
    errors: list[str] = field(default_factory=list)
    oversized: list[Domain] = field(default_factory=list)


class _UnionFind:
    def __init__(self, n: int) -> None:
        self._parent = list(range(n))

    def find(self, x: int) -> int:
        while self._parent[x] != x:
            self._parent[x] = self._parent[self._parent[x]]
            x = self._parent[x]
        return x

    def union(self, a: int, b: int) -> None:
        ra, rb = self.find(a), self.find(b)
        if ra != rb:
            self._parent[rb] = ra


class DomainPartitioner:
    """Connected components of the seat graph; each component is one domain."""

    def __init__(self, graph: PhysicalSeatGraph) -> None:
        self.graph = graph
        self._components = self._connected_components()

    def _connected_components(self) -> list[list[int]]:
        uf = _UnionFind(len(self.graph.nodes))
        for i, j in self.graph.edges:
            uf.union(i, j)
        groups: dict[int, list[int]] = {}
        for idx in range(len(self.graph.nodes)):
            root = uf.find(idx)
            groups.setdefault(root, []).append(idx)
        for seats in groups.values():
            seats.sort()
        return list(groups.values())

    def domains(self) -> list[Domain]:
        domains: list[Domain] = []
        for i, seats in enumerate(self._components):
            hall_ids: list[str] = []
            building_ids: list[str] = []
            for s in seats:
                node = self.graph.nodes[s]
                if node.hall_id not in hall_ids:
                    hall_ids.append(node.hall_id)
                if node.building is not None and node.building not in building_ids:
                    building_ids.append(node.building)
            domains.append(
                Domain(
                    domain_id=f"component-{i:02d}",
                    seat_indices=seats,
                    hall_ids=hall_ids,
                    building_ids=building_ids,
                )
            )
        return domains

    def verify_partition_invariants(self) -> list[str]:
        """§38 invariants: every seat/edge in exactly one domain; edges never cross domains."""
        errors: list[str] = []
        seat_of_domain: dict[int, str] = {}
        for domain in self.domains():
            for s in domain.seat_indices:
                if s in seat_of_domain:
                    errors.append(f"seat index {s} appears in multiple domains")
                seat_of_domain[s] = domain.domain_id
        for s in range(len(self.graph.nodes)):
            if s not in seat_of_domain:
                errors.append(f"seat index {s} missing from every domain")
        for i, j in self.graph.edges:
            if seat_of_domain.get(i) != seat_of_domain.get(j):
                errors.append(
                    f"cross-domain adjacency edge ({i}, {j}): "
                    f"{seat_of_domain.get(i)} <-> {seat_of_domain.get(j)}"
                )
        return errors

def allocate_candidates_to_domains(
    candidates: list[Candidate],
    domains: list[Domain],
    department_key: Callable[[Candidate], str] = lambda c: c.department,
) -> list[Domain]:
    """Deterministic balanced candidate allocation for benchmarking/scheduling.

    Candidates of each department (in register order) are distributed
    round-robin across domains that still have free seat capacity. This keeps
    every domain's department mix balanced and its count within its seat count.
    """
    seat_counts = [d.seat_count for d in domains]
    filled = [0] * len(domains)
    assigned: list[list[str]] = [[] for _ in domains]

    by_department: dict[str, list[Candidate]] = {}
    for c in candidates:
        by_department.setdefault(department_key(c), []).append(c)

    domain_order = list(range(len(domains)))
    for dept in sorted(by_department):
        cursor = 0
        for candidate in sorted(by_department[dept], key=lambda c: (c.registerNumber, c.id)):
            for _ in range(len(domains)):
                d = domain_order[cursor % len(domain_order)]
                cursor += 1
                if filled[d] < seat_counts[d]:
                    assigned[d].append(candidate.id)
                    filled[d] += 1
                    break
            else:
                raise ValueError(f"no domain has free seat capacity for candidate {candidate.id}")

    for d, domain in enumerate(domains):
        domain.candidate_ids = assigned[d]
    return domains


def partition_request(
    candidates: list[Candidate],
    graph: PhysicalSeatGraph,
    max_domain_candidates: int = MAX_DOMAIN_CANDIDATES_DEFAULT,
) -> PartitionResult:
    """Full §38 pipeline: build domains, verify invariants, allocate candidates, ceiling."""
    partitioner = DomainPartitioner(graph)
    domains = partitioner.domains()
    errors = partitioner.verify_partition_invariants()

    try:
        allocate_candidates_to_domains(candidates, domains)
    except ValueError as exc:
        errors.append(f"{ERR_INVALID_DOMAIN_ASSIGNMENT}: {exc}")

    oversized = [d for d in domains if d.candidate_count > max_domain_candidates]
    for d in oversized:
        errors.append(
            f"{ERR_GRAPH_TOPOLOGY_OVERSIZED_COMPONENT}: domain {d.domain_id} "
            f"candidates={d.candidate_count} exceeds ceiling {max_domain_candidates}"
        )
    return PartitionResult(domains=domains, errors=errors, oversized=oversized)


def topology_anomaly_evidence(
    domain: Domain,
    max_domain_candidates: int = MAX_DOMAIN_CANDIDATES_DEFAULT,
    graph: Optional[PhysicalSeatGraph] = None,
) -> dict:
    """§25 — structured evidence for an oversized domain (recorded, solver must not run)."""
    bridge_edges: list[dict] = []
    if graph is not None:
        nodes = graph.nodes
        for i, j in graph.edges:
            if i in domain.seat_indices and j in domain.seat_indices:
                bridge_edges.append(
                    {
                        "from": nodes[i].seat_id,
                        "fromHall": nodes[i].hall_number,
                        "to": nodes[j].seat_id,
                        "toHall": nodes[j].hall_number,
                    }
                )
    return {
        "domain_id": domain.domain_id,
        "candidate_count": domain.candidate_count,
        "seat_count": domain.seat_count,
        "ceiling": max_domain_candidates,
        "hall_ids": domain.hall_ids,
        "building_ids": domain.building_ids,
        "adjacency_edges_within_domain": bridge_edges,
    }